src/Induction/NoInfiniteDescent.agda

```diff
@@ -0,0 +1,221 @@
+open import Data.Product.Base using (_,_; proj₁; ∃-syntax; _×_)
+open import Function.Base using (_∘_)
+open import Induction.WellFounded
+  using (WellFounded; Acc; acc; acc-inverse; module Some)
+open import Level using (Level)
+open import Relation.Binary.Core using (Rel)
+open import Relation.Binary.Construct.Closure.Transitive
+open import Relation.Binary.PropositionalEquality.Core
+open import Relation.Nullary.Negation.Core as Negation using (¬_)
+open import Relation.Unary
+  using (Pred; ∁; _∩_; _⇒_; Universal; IUniversal; Stable)
+
+private
+  variable
+    a r ℓ : Level
+    A : Set a
+
+------------------------------------------------------------------------
+-- Definitions
+
+module InfiniteDescent (_<_ : Rel A r) (P : Pred A ℓ)  where
+
+    DescentAt : Pred A _
+    DescentAt x = P x → ∃[ y ] y < x × P y
+
+    Acc⇒Descent : Set _
+    Acc⇒Descent = ∀[ Acc _<_ ⇒ DescentAt ]
+
+    Descent : Set _
+    Descent = ∀[ DescentAt ]
+
+    InfiniteDescendingSequence : (f : ℕ → A) → Set _
+    InfiniteDescendingSequence f = ∀ n → f (suc n) < f n
+
+    InfiniteSequence_From_ : (f : ℕ → A) → Pred A _
+    InfiniteSequence f From x = f zero ≡ x × InfiniteDescendingSequence f
+
+    InfiniteDescentAt : Pred A _
+    InfiniteDescentAt x = P x → ∃[ f ] Π[ P ∘ f ] × InfiniteSequence f From x
+
+    Acc⇒InfiniteDescent : Set _
+    Acc⇒InfiniteDescent = ∀[ Acc _<_ ⇒ InfiniteDescentAt ]
+
+    InfiniteDescent : Set _
+    InfiniteDescent = ∀[ InfiniteDescentAt ]
+
+------------------------------------------------------------------------
+-- Basic lemmas: assume unrestricted descent
+
+    module Lemmas (descent : Descent) where
+
+      acc⇒infiniteDescent : Acc⇒InfiniteDescent
+      acc⇒infiniteDescent {x} = Some.wfRec InfiniteDescentAt rec x
+        where
+        rec : _
+        rec y rec[y] py
+          with z , z<y , pz ← descent py
+          with g , Π[P∘g] , g0≡z , g<P ← rec[y] z<y pz
+             = f , Π[P∘f] , f0≡y , f<P
+          where
+          f : ℕ → A
+          f zero = y
+          f (suc n) = g n
+          f0≡y : f zero ≡ y
+          f0≡y = refl
+          f<P : ∀ n → f (suc n) < f n
+          f<P zero rewrite g0≡z = z<y
+          f<P (suc n)           = g<P n
+          Π[P∘f] : Π[ P ∘ f ]
+          Π[P∘f] zero rewrite g0≡z = py
+          Π[P∘f] (suc n)           = Π[P∘g] n
+
+      wf⇒infiniteDescent : WellFounded _<_ → InfiniteDescent
+      wf⇒infiniteDescent wf = acc⇒infiniteDescent (wf _)
+
+      acc⇒noInfiniteDescent : ∀[ Acc _<_ ⇒ ¬_ ∘ P ]
+      acc⇒noInfiniteDescent {x} = Some.wfRec (¬_ ∘ P) rec x
+        where
+        rec : _
+        rec y rec[y] py = let z , z<y , pz = descent py in rec[y] z<y pz
+
+      wf⇒noInfiniteDescent : WellFounded _<_ → ∀[ ¬_ ∘ P ]
+      wf⇒noInfiniteDescent wf = acc⇒noInfiniteDescent (wf _)
+
+------------------------------------------------------------------------
+-- Corollaries: assume descent only for Acc _<_ elements
+
+module _ {_<_ : Rel A r} (P : Pred A ℓ) where
```

Comment:
Given the amount of module indirection, it is very difficult to work out whether `_<_` can truly be implicit for all the lemmas in this module... Have you checked them all?

Reply:
Er, ... no, I can't honestly say I have. I think I've been working on gut instinct that Agda can infer `_<_` from instances of `Descent`/`DescentAt` but I may very well be mistaken in those instincts. Good catch!

Reply:
So... I adopted the solution that all appeals to `_<_` would be explicit, at the potential cost of having to supply `_` arguments as module/definition parameters... hope this fits the bill!